### PR TITLE
feat: add macro crate

### LINKED ISSUE #3933: Rewrite `format_hardhat_call` as a proc macro
Rewrite `format_hardhat_call` as a proc macro. Adding any new console log overloads as is would be a waste of time
https://github.com/foundry-rs/foundry/blob/60b19191ce1f7eb9fac5dcccf46442561d922bef/evm/src/executor/abi/fmt.rs#L172

Ref #3932

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Fixes #3933

## Solution

add `macro` crate
rename private trait `FormatValue` to public `ConsoleFmt` (similar to `UIFmt`)
add derive macro `ConsoleFmt` that derives the trait which dispatches struct and enum `calls` to generic format function

### CODE REVIEW

**Comment by DaniPopes** [comment]
not sure how to debug the CI error:

`/tmp/nextest-archive-WNEkxU/target/debug/deps/foundry_macros_impl-d13a0a5a9740e9f5: error while loading shared libraries: libtest-5678a62cd2e949f2.so: cannot open shared object file: No such file or directory
`

File: .github/workflows/cross-platform.yml [hunk 1]
```diff
@@ -38,7 +38,7 @@ jobs:
 
         steps:
             - name: Checkout sources
-              uses: actions/checkout@v2
+              uses: actions/checkout@v3
```
Comment (mattsse) [inline .github/workflows/cross-platform.yml:41]: are these changes relevant to the PR?

if not please extract and submit separately

File: macros/impl/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+[package]
+name = "foundry-macros-impl"
```
Comment (mattsse) [inline macros/impl/Cargo.toml:2]: how about renaming this to `-derive`? instead of `-impl`?

**Review by mattsse — CHANGES_REQUESTED** [review]
tysm!

solving the console.log varargs issue via derive will be less cumbersome to maintain.

unsure about the crate names though, wdyt @gakonst ?

File: macros/impl/Cargo.toml (hunk 2 above)
Comment (DaniPopes) [reply macros/impl/Cargo.toml:2]: I was thinking about that, but then that also implies new crates for attributes and normal proc macros, so I don't know

File: .github/workflows/cross-platform.yml (hunk 1 above)
Comment (DaniPopes) [reply .github/workflows/cross-platform.yml:41]: was trying to fix CI error but don't know how, it runs fine if I run the same commands locally (with nextest archive etc)

File: macros/impl/Cargo.toml (hunk 2 above)
Comment (mattsse) [reply macros/impl/Cargo.toml:2]: I see, that makes sense.

File: .github/workflows/cross-platform.yml (hunk 1 above)
Comment (mattsse) [reply .github/workflows/cross-platform.yml:41]: please revert if not related and/or submit separately.

File: macros/impl/src/console_fmt.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            Fields::Unit => return quote!(),
+        };
+
+        let fields: Punctuated<Ident, Token![,]> = fields
+            .enumerate()
+            .map(|(i, field)| field.ident.unwrap_or_else(|| format_ident!("__var_{i}")))
+            .collect();
+
+        if fields.len() != 1 {
+            unimplemented!("Enum variant with not 1 field")
```
Comment (rkrasiuk) [inline macros/impl/src/console_fmt.rs:96]: nit
```suggestion
            unimplemented!("Enum variant with more than 1 field")
```

**Review by rkrasiuk — APPROVED** [review]
@DaniPopes well done!

**Comment by rkrasiuk** [comment]
@DaniPopes the test failures seem to be related, did u have a chance to look into them?

**Comment by DaniPopes** [comment]
So what was going on here is that the `proc_macro` crate (`foundry-macros-impl`) was being compiled to an executable that was dynamically linked to the runner's local `libstd` and `libtest`, which are found at `<sysroot>/lib` (`rustc --print sysroot`). These libraries were either 1. not present, or, if present, 2. not found by `ld` in the test runner(s).

To patch this we have to first install the same Rust toolchain as the build runner (they must be the same version so that the libraries' hashes match), then set the `LD_LIBRARY_PATH` env variable to the `<sysroot>/lib` path so `ld` can find them.

Related to: nextest-rs/nextest#267

**Comment by gakonst** [comment]
@mattsse any ideas on this edge case around cargo nexttest? I guess its "fine" but I don't love it. We could always try to do the fix upstream, but wonder if there's a simpler workaround.

**Review by gakonst — APPROVED** [review]

**Review by mattsse — APPROVED** [review]
lgtm

let's track the nextest limitation separately, seems kinda off...
